### DISCUSSION TITLE
Docker headless no AVX compile but Illegal instruction

### DISCUSSION THREAD

**Discussion by jerome-godbout-lychens** [isl-org/Open3D#7010] [Q&A]
Hi, 
Since I was getting illegal instruction when installing open3d from pip:

```
FROM python:3.10-slim
ENV DEBIAN_FRONTEND=noninteractive
ENV PYTHONUNBUFFERED=1
....
RUN pip3 install --upgrade pip
RUN pip3 install -y open3d
```

Inside the docker image if I try to use the open3d I get illegal instruction:

```
# python
Python 3.10.15 (main, Sep 27 2024, 06:07:40) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import open3d
Illegal instruction
```

I realized I was not having AVX cpu in my docker env where it run.  So I am trying to run open3d with visualization.rendering.OffscreenRenderer in a headless, non AVX CPU environnement. I'm trying to compile this way:

This is inefficient Docker code, just to debug ans test until I get it right

```
FROM python:3.10-slim
ENV DEBIAN_FRONTEND=noninteractive
ENV PYTHONUNBUFFERED=1
....
RUN git clone --recursive https://github.com/isl-org/Open3D.git --branch v0.18.0 --single-branch /Open3D
WORKDIR /Open3D
RUN mkdir build 
RUN apt-get install -y libegl1 \
    libgl1 \
    libgomp1 \
    libxinerama-dev \
    libxcursor-dev \
    libglfw3-dev \
    libglew-dev \
    libx11-dev \
    libxrandr-dev \
    libxi-dev \
    libudev-dev \
    libpng-dev \
    libjpeg-dev \
    libtiff-dev \
    libopenexr-dev 

RUN cd build && cmake -DBUILD_PYTHON_MODULE=ON -DBUILD_SHARED_LIBS=ON -DPYTHON_EXECUTABLE=/usr/local/bin/python -DBUILD_GUI=OFF -DOPEN3D_USE_AVX=OFF .. 
RUN cd build && make
RUN cd build && make install-pip-package 
RUN cd build && make python-package
RUN cd build && make pip-package
RUN cd build && make install
RUN cd /Open3D/build/lib/python_package && pip install .
```

But I cannot get away from the illegale instruction when importing:
```
# python
Python 3.10.15 (main, Sep 27 2024, 06:07:40) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import open3d
Illegal instruction
```

Anyway to known what is the problem? Any other things that must be turn off?

**Comment by ssheorey** [comment]
Try building with `-DBUILD_ISPC_MODULE=OFF`. ISPC provides vectorization (SIMD) support.

What is your HW platform / processor? Here are the SIMD instructions that we build - it should work on your system if any one is present. (SSE2 should be sufficient).

```
Using ISPC instruction sets: sse2-i32x4;sse4-i32x4;avx1-i32x8;avx2-i32x8;avx512knl-i32x16;avx512skx-i32x16
```

**Reply by jerome-godbout-lychens** [reply]
Still not enough, I still get illegal instructions.

**Comment by jerome-godbout-lychens** [comment]
I'm running this inside a docker, my local machine is Apple M3 Pro CPU (arm). Would be nice to have a runtime detection if the instruction set is present to use the right vectoriel instructions. So running the docker on x64 machine would still have the speed boost but no need for another image for non x64 native runner of a docker file.

**Comment by jerome-godbout-lychens** [comment]
I end up doing one docker for the open3d part and another docker for colmap (colmap is not arm friendly at this time). This avoid trying to run the open3d on rosetta with arm cpu under MacOS. This solve my problem with my docker images for now. But would be nice to have the avx instructions being only used when they are available and use standard (slower) instructions when not. This could be detected at module load easily with: `cat /proc/cpuinfo | grep avx` for Linux platform with x86/x64 arch. It's a bit sad the AVX instructions are not emulated into the docker when emulating x86/x64 platform with Rosetta on MacOS arm.
